src/libostree/ostree-repo-static-delta-core.c

```diff
@@ -895,3 +982,168 @@ _ostree_repo_static_delta_dump (OstreeRepo                    *self,
 
   return TRUE;
 }
+
+static void
+null_or_ptr_array_unref (GPtrArray *array)
+{
+  if (array != NULL)
+    g_ptr_array_unref (array);
+}
+
+static gboolean
+file_has_content (OstreeRepo   *repo,
+                  const char   *subpath,
+                  GBytes       *data,
+                  GCancellable *cancellable)
+{
+  struct stat stbuf;
+  glnx_autofd int existing_fd = -1;
+
+  if (!glnx_fstatat (repo->repo_dir_fd, subpath, &stbuf, 0, NULL))
+    return FALSE;
+
+  if (stbuf.st_size != g_bytes_get_size (data))
+    return FALSE;
+
+  if (!glnx_openat_rdonly (repo->repo_dir_fd, subpath, TRUE, &existing_fd, NULL))
+    return FALSE;
+
+  g_autoptr(GBytes) existing_data = glnx_fd_readall_bytes (existing_fd, cancellable, NULL);
+  if (existing_data == NULL)
+    return FALSE;
+
+  return g_bytes_equal (existing_data, data);
+}
+
+gboolean
+_ostree_repo_static_delta_reindex (OstreeRepo                 *repo,
+                                   const char                 *opt_to_commit,
+                                   GCancellable               *cancellable,
+                                   GError                    **error)
+{
```

Comment:
Even though the summary path does take an exclusive lock, I think you should at least take a shared lock here to ensure you're not racing with another process pruning deltas. An exclusive lock would (I believe) also ensure that another process isn't creating deltas while indexing. With the CLI you added, this can also be called outside of the summary generation path.

Reply:
I think a shared lock makes sense, to avoid a parallel prune removing commits that the generated indexes will refer too. However, I think an exclusive lock is overkill (and unnecessarily deadlock risky), because at most a parallel call that was generated before the reindex is finished will be missing. 

If you want to guarantee that a newly generated delta is index you have to manually ensure reindex is called after the delta is fully created. An exclusive lock on the indexing will not help you with that, all it will do is delay the start of delta creation if a reindex is running.